### ISSUE TITLE
State result parsing fails with error when there is a job failure

### ISSUE THREAD

**Issue by paclat** [saltstack/salt#32108]
### Description of Issue/Question
We had a situation in volume testing when Salt would incorrectly think that a job was done.  It would then try to run the state again but it would fail with an error message.  When it did so we would get a cryptic error message:
```
2016-03-21 19:15:45,211 [salt.state                               ][ERROR   ][1003][139775351428864] An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 1564, in call
    **cdata['kwargs'])
  File "/usr/lib64/python2.6/contextlib.py", line 34, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/lib/python2.6/site-packages/salt/utils/context.py", line 42, in func_globals_inject
    yield
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 1564, in call
    **cdata['kwargs'])
  File "/usr/lib/python2.6/site-packages/salt/states/saltmod.py", line 267, in state
    for state_item in six.itervalues(m_ret):
  File "/usr/lib/python2.6/site-packages/salt/ext/six.py", line 569, in itervalues
    return iter(d.itervalues(**kw))
AttributeError: 'list' object has no attribute 'itervalues'
```
### Setup
(Please provide relevant configs and/or SLS files (Be sure to remove sensitive info).)

### Steps to Reproduce Issue
Reproducing this will involve explicitly trying to run two states at the same time..

### Versions Report
```
           Salt: 2015.5.9
         Python: 2.6.6 (r266:84292, Jul 23 2015, 05:13:40)
         Jinja2: unknown
       M2Crypto: 0.20.2
 msgpack-python: 0.4.6
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.11
          ioflo: Not Installed
          PyZMQ: 14.5.0
           RAET: Not Installed
            ZMQ: 4.0.5
           Mako: Not Installed
        Tornado: Not Installed
        timelib: Not Installed
       dateutil: Not Installed
```

**Comment by paclat** [comment]
The fix that worked for us when patched was to change the code in question from ..

`
      for state_item in six.itervalues(m_ret):
            if state_item['changes']:
                changes[minion] = m_ret
                break
        else:
            no_change.add(minion)
`

to 

`
       try:
            for state_item in six.itervalues(m_ret):
                if 'changes' in state_item and state_item['changes']:
                    changes[minion] = m_ret
                    break
            else:
                no_change.add(minion)
        except:
            log.error("m_ret did not have changes %s %s", type(m_ret), m_ret)
            no_change.add(minion)
`

**Comment by jfindlay** [comment]
@paclat, thanks for reporting.  Your change seems good to me except that I would only `except AttributeError` so that this change does not dispose of other errors that may arise.  You should also seek the advice and consent of @rallytime or @cachedout.

**Comment by paclat** [comment]
Thanks @jfindlay .

Created pull request that includes your suggestion of AttributeError

https://github.com/saltstack/salt/pull/32141

Hopefully I did this correctly...

@rallytime or @cachedout  please review.

**Comment by cachedout** [comment]
Fixed by #32141. Closing.
